modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java

```diff
@@ -166,45 +171,113 @@ private void ensureOpen() {
     }
 
     private void reopenStreamOrFail(IOException e) throws IOException {
-        if (purpose == OperationPurpose.REPOSITORY_ANALYSIS) {
-            logger.warn(() -> format("""
-                failed reading [%s/%s] at offset [%s]""", blobStore.bucket(), blobKey, start + currentOffset), e);
-            throw e;
+        maybeRecordAndLogForRetry("reading", e);
+        maybeAbort(currentStream);
+        IOUtils.closeWhileHandlingException(currentStream);
+
+        maybeDelay();
+        openStreamWithRetry();
+    }
+
+    private <T extends Exception> void maybeRecordAndLogForRetry(String action, T e) throws T {
+        if (shouldRetry() == false) {
+            final var finalException = addSuppressedExceptions(e);
+            logForRetry(Level.WARN, action, finalException);
+            throw finalException;
         }
 
-        final int maxAttempts = blobStore.getMaxRetries() + 1;
+        attempt += 1;
```

Comment:
Would it not be more intuitive to increment this after `logForRetry` and avoid subtracting 1 from attempt there?

Reply:
This is related to the above comment. I had to increment it first because `attempt == 1` was used to signal failure without retries. It was confusing. I have now made them separate and therefore `attempt` is now incremented after the logging.